lib/hmmlearn/hmm.py

```diff
@@ -1011,69 +995,76 @@ def _do_mstep(self, stats):
 
         # Maximizing covariances
         if 'c' in self.params:
+            lambdas, mus = self.means_weight, self.means_prior
             centered_means = self.means_ - mus
 
             def outer_f(x):  # Outer product over features.
                 return x[..., :, None] * x[..., None, :]
 
             if self.covariance_type == 'full':
-                centered_dots = outer_f(centered)
                 centered_means_dots = outer_f(centered_means)
 
                 psis_t = np.transpose(self.covars_prior, axes=(0, 1, 3, 2))
                 nus = self.covars_weight
 
-                c_n = (
-                    np.einsum('ijk,ijklm->jklm', post_comp_mix, centered_dots)
-                    + psis_t
-                    + lambdas[:, :, None, None] * centered_means_dots
-                )
+                c_n = psis_t + lambdas[:, :, None, None] * centered_means_dots
+                for post_comp_mix, samples in zip(stats['post_comp_mix'],
+                                                  stats['samples']):
+                    centered = samples[:, None, None, :] - means_before_update
+                    centered_dots = outer_f(centered)
+                    c_n += np.einsum('ijk,ijklm->jklm', post_comp_mix,
+                                     centered_dots)
                 c_d = (
                     stats['post_mix_sum'] + 1 + nus + nf + 1
                 )[:, :, None, None]
 
             elif self.covariance_type == 'diag':
-                centered2 = centered ** 2
-                centered_means2 = centered_means ** 2
-
                 alphas = self.covars_prior
                 betas = self.covars_weight
+                centered_means2 = centered_means ** 2
+
+                c_n = lambdas[:, :, None] * centered_means2 + 2 * betas
+                for post_comp_mix, samples in zip(stats['post_comp_mix'],
+                                                  stats['samples']):
+                    centered = samples[:, None, None, :] - means_before_update
+                    centered2 = np.square(centered, out=centered)  # reuse
+                    c_n += np.einsum('ijk,ijkl->jkl', post_comp_mix, centered2)
```

Comment:
Comment only, out of scope for this PR:

If the goal is to reduce peak memory use, it seems that there's still some waste here as we construct a huge intermediate `centered` value of shape `(n_samples_in_this_sequence, n_components, n_mix, n_features` and then do a single pass over it, immediately reducing away the `n_samples_in_this_sequence` dimensions in the `einsum`.

I reckon the calculation here is is equivalent to this naive bunch of C for loops, that doesn't need to allocate any memory for a huge `centered` array at all:

```
int i, j, k, l;
double acc, centered;

for (j = 0; j < n_components; ++j) {
    for (k = 0; k < n_mix; ++k) {
        for (l = 0; l < n_features; ++l) {
            acc = 0.0;
            for (i = 0; i < n_samples_in_this_sequence; ++i) {
                centered = samples[i, l] - means_before_update[j, k, l];
                acc += post_comp_mix[i, j, k] *  centered * centered;
            }
            c_n[j, k, l] += acc;
        }
    }
}
```

We could implement the logic by writing it in Cython, at some cost to maintainability.  I am not sure there is a clean way to express it in fast pure numpy code, as `einsum` doesn't let us use subtraction in a big tensor expression.

I'm not sure if similar idea could be used to rewrite the other cases of covariance_type.

Reply:
I avoid any Cython optimizations - they become hard to maintain in a year, especially for outsiders.
Feel free to create a dedicated PR.